File: src/main/java/xyz/phanta/cbtweaker/structure/block/StructureBlockMatcher.java
package xyz.phanta.cbtweaker.structure.block;

import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import xyz.phanta.cbtweaker.util.Direction;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public interface StructureBlockMatcher {

    @Nullable
    StructureBlockMatch matchBlock(World world, BlockPos pos, Direction dir);

    List<StructureBlockVisualization> getVisualization();

    class MatcherSet implements StructureBlockMatcher {

        public static MatcherSet of(StructureBlockMatcher... matchers) {
            MatcherSet matcherSet = new MatcherSet();
            matcherSet.addAll(Arrays.asList(matchers));
            return matcherSet;
        }

        private final List<StructureBlockMatcher> matchers = new ArrayList<>();
        @Nullable
        private List<StructureBlockVisualization> visList;

        public void add(StructureBlockMatcher matcher) {
            matchers.add(matcher);
        }

        public void addAll(Collection<StructureBlockMatcher> matchers) {
            this.matchers.addAll(matchers);
        }

        @Nullable
        @Override
        public StructureBlockMatch matchBlock(World world, BlockPos pos, Direction dir) {
            for (StructureBlockMatcher matcher : matchers) {
                StructureBlockMatch match = matcher.matchBlock(world, pos, dir);
                if (match != null) {
                    return match;
                }
            }
            return null;
        }

        @Override
        public List<StructureBlockVisualization> getVisualization() {
            if (visList == null) {
                visList = matchers.stream().flatMap(m -> m.getVisualization().stream()).collect(Collectors.toList());
            }
            return visList;
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/structure/block/StructureBlockMatcher.kt
package st.evening.mc.cbtweaker.structure.block

import net.minecraft.util.Rotation
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import st.evening.mc.cbtweaker.util.Identifiable
import st.evening.mc.prelude.api.data.tjson.JsonPath
import st.evening.mc.prelude.api.data.tjson.TJson

interface StructureBlockMatcher {
    val visualization: List<StructureBlockVisualization>

    fun matchBlock(world: World, pos: BlockPos, rotation: Rotation): StructureBlockMatch?

    object NothingMatcher : StructureBlockMatcher {
        override val visualization: List<StructureBlockVisualization>
            get() = emptyList()

        override fun matchBlock(world: World, pos: BlockPos, rotation: Rotation): StructureBlockMatch? = null
    }

    class MatcherSet(private val matchers: List<StructureBlockMatcher>) : StructureBlockMatcher {
        constructor(vararg matchers: StructureBlockMatcher) : this(matchers.asList())

        override val visualization: List<StructureBlockVisualization> by lazy {
            matchers.flatMap { it.visualization }
        }

        override fun matchBlock(world: World, pos: BlockPos, rotation: Rotation): StructureBlockMatch? =
            matchers.firstNotNullOfOrNull { it.matchBlock(world, pos, rotation) }
    }
}

fun List<StructureBlockMatcher>.joinMatchers(): StructureBlockMatcher = when (size) {
    0 -> StructureBlockMatcher.NothingMatcher
    1 -> this[0]
    else -> StructureBlockMatcher.MatcherSet(this)
}

interface StructureBlockMatcherType : Identifiable {
    context(_: JsonPath)
    fun loadMatcher(dto: TJson.Object): StructureBlockMatcher
}
